File: FoxCloud-node/src/main/kotlin/io/github/thecguygithub/node/command/CommandProvider.java
package io.github.thecguygithub.node.command;

import io.github.thecguygithub.api.command.CommandInfo;
import io.github.thecguygithub.node.command.source.CommandSource;
import lombok.NonNull;
import org.jetbrains.annotations.Nullable;
import org.jetbrains.annotations.UnmodifiableView;

import java.util.Collection;
import java.util.List;
import java.util.concurrent.CompletableFuture;

public interface CommandProvider {

    @NonNull
    List<String> suggest(@NonNull CommandSource source, @NonNull String input);

    CompletableFuture<?> execute(@NonNull CommandSource source, @NonNull String input);

    void register(@NonNull Class<?> commandClass);

    void register(@NonNull Object command);

    @Nullable
    CommandInfo command(@NonNull String name);

    @UnmodifiableView
    @NonNull
    Collection<CommandInfo> commands();

}

File: VulpesCloud-node/src/main/kotlin/de/vulpescloud/node/command/CommandProvider.kt
package de.vulpescloud.node.command

import de.vulpescloud.node.command.impl.CommandInfo
import org.incendo.cloud.CommandManager
import org.incendo.cloud.execution.CommandResult
import java.util.concurrent.CompletableFuture

interface CommandProvider {

    val commandManager: CommandManager<CommandSource>

    fun initialize()

    fun suggest(source: CommandSource, input: String): List<String>

    fun execute(source: CommandSource, input: String): CompletableFuture<CommandResult<CommandSource>>

    fun register(command: Any)

    fun command(name: String): CommandInfo?

    fun commands(): MutableCollection<CommandInfo>?

    fun commandUsageOfRoot(root: String): List<String>
}